tools/src/main/java/org/apache/kafka/trogdor/workload/ExternalCommandWorker.java

```diff
@@ -0,0 +1,227 @@
+
+    private ScheduledExecutorService executor;
+
+    private Process worker;
+
+    private String errMsg = "";
+
+    private WorkerStatusTracker status;
+
+    private KafkaFutureImpl<String> doneFuture;
+
+    public ExternalCommandWorker(String id, ExternalCommandSpec spec) {
+        this.id = id;
+        this.spec = spec;
+    }
+
+    @Override
+    public void start(Platform platform, WorkerStatusTracker status,
+                      KafkaFutureImpl<String> doneFuture) throws Exception {
+        if (!running.compareAndSet(false, true)) {
+            throw new IllegalStateException("ExternalCommandWorker is already running.");
+        }
+        log.info("{}: Activating ExternalCommandWorker with {}", id, spec);
+        this.executor = Executors.newScheduledThreadPool(2,
+            ThreadUtils.createThreadFactory("ExternalCommandWorkerThread%d", false));
+        this.status = status;
+        this.doneFuture = doneFuture;
+        executor.submit(new Prepare());
+    }
+
+    public class Prepare implements Runnable {
+        @Override
+        public void run() {
+            try {
+                ArrayList<String> command = new ArrayList<String>(spec.command());
+                command.add("--spec");
+                command.add(spec.toString());
+                log.info("ExternalCommandWorker executes command: {}", command);
+                ProcessBuilder workerBuilder = new ProcessBuilder(command);
+                worker = workerBuilder.start();
+                Future updater = executor.submit(new StatusUpdater());
+                Future errorUpdater = executor.submit(new ErrorUpdater());
+                worker.waitFor();
+                updater.get();
+                errorUpdater.get();
+                int workerExitValue = worker.exitValue();
+                log.info("ExternalWorker finished with the exit value: {}.", workerExitValue);
+                if (workerExitValue != 0 && errMsg.isEmpty()) {
+                    errMsg = "ExternalWorker exited with error code " + worker.exitValue();
+                }
+                log.info("StatusUpdater terminated.");
+                doneFuture.complete(errMsg);
+            } catch (IOException e) {
+                log.info("Stdout of ExternalWorker is closed.");
+            } catch (InterruptedException e) {
+                log.info("StatusUpdadter interrupted.");
+            } catch (Exception e) {
+                WorkerUtils.abort(log, "Prepare", e, doneFuture);
+            }
+        }
+    }
+
+    public class ErrorUpdater implements Runnable {
+        BufferedReader br;
+        ErrorUpdater() {
+            br = new BufferedReader(new InputStreamReader(worker.getErrorStream(), StandardCharsets.UTF_8));
+        }
+        @Override
+        public void run() {
+            try {
+                String line;
+                while ((line = br.readLine()) != null) {
+                    log.error("ExternalWorker (stderr):{}", line);
```

Comment:
Can we put the worker ID into the prefix here?  We may be running multiple workers at once.